### ISSUE TITLE
Running map is slow compared to train or test

### ISSUE THREAD

**Issue by jveitchmichaelis** [AlexeyAB/darknet#2137]
I'm running the `-map` flag during a training run. I'm using yolov3 with some 640x512 images, single class. On my 1080ti I can process a batch of 64 images in 8 seconds (GPU but no CUDNN).

Whenever the map calculation runs, it seems to take ages to get the detection results and my GPU utilisation is poor. It takes over a second per image (order of magnitude slower). I tried increasing the number of threads in the code, but that didn't make a difference. This is also the case if I run the `./darknet detector map` as a standalone command.

I can run `detector test` on an image in the validation set in about 130msecs using the same model/weights etc.

Do I have a configuration problem somewhere? I'm sure I've had it running much faster in the past.

One suspicion I have is that I modified the code to do a full MAP calculation frequently - every 200 iterations - as I found that my model loss was decreasing extremely quickly. I guess if the early model predicts a huge number of incorrect bounding boxes, the processing time will be longer (and weights from later epochs will run faster). Might be worth profiling to see if any improvements can be made to the map calculation though, even with 500 boxes, what's the bottleneck? My CPU is also only maxing a single core.

EDIT:

Just ran this through callgrind - and yeah it's probably the non-max suppression that's causing the slowdown.

![non-max suppression profile](https://i.imgur.com/tKtmoGa.png)

**Comment by AlexeyAB** [comment]
@jveitchmichaelis Hi,

* Do you use the latest version of Darknet? There was solved several bugs for `-map` during training.

* By default it calculates mAP for each 4 epochs `= iterations / images_int_train_txt`. And not earlier than `1000` iterations or `3 * burn_in` iterations, since it doesn't make sense, because the weight is too bad yet: https://github.com/AlexeyAB/darknet/blob/527578744b46666fb5cd42393bf9e1fa9af126ee/src/detector.c#L249-L250

----


> One suspicion I have is that I modified the code to do a full MAP calculation frequently - every 200 iterations - as I found that my model loss was decreasing extremely quickly.

It looks like you removed this condition for calculation mAP not earlier than `1000` iterations or `3 * burn_in` iterations. So you tried to check mAP for yolo_200.weights where is too much false positives, so do_nms_sort is stugled.
Or you use old code of Darknet with bug in `-map` during training.

Try to download new code from GitHub and undo your changes.
Currently I don't meet with performance issues for `-map` during training on different machines and OSs.

**Comment by jveitchmichaelis** [comment]
Hi @AlexeyAB,

Thanks, I think it's just NMS being crushed for a huge number of boxes. I found (for my dataset) after 500 iterations it's fine, but I wanted to capture the initial changes in mAP and 1000 was too late. I think as I'm training on a single fairly easy class the network is converging very quickly.

One other issue though, currently mAP is ignored unless you compile with opencv. This isn't necessary. I don't plot the results, I just direct stdout to a file and parse that later. I modified the code to something like this which works fine:

```
int draw_precision = 0;
int calc_map_for_each = 4 * train_images_num / (net.batch * net.subdivisions);  // calculate mAP for each 4 Epochs
if (calc_map && (i >= (iter_map + calc_map_for_each) || i == net.max_batches) && i >= net.burn_in && i >= 1000) {
    if (l.random) {
        printf("Resizing to initial size: %d x %d \n", init_w, init_h);
        args.w = init_w;
        args.h = init_h;
        pthread_join(load_thread, 0);
        train = buffer;
        load_thread = load_data(args);
        int k;
        for (k = 0; k < ngpus; ++k) {
            resize_network(nets + k, init_w, init_h);
        }
        net = nets[0];
    }

    // combine Training and Validation networks
    network net_combined = combine_train_valid_networks(net, net_map);

    iter_map = i;
    mean_average_precision = validate_detector_map(datacfg, cfgfile, weightfile, 0.25, 0.5, &net_combined);
    printf("\n mean_average_precision = %f \n", mean_average_precision);
    draw_precision = 1;

    #ifdef OPENCV
    if (!dont_show) {
        draw_train_loss(img, img_size, avg_loss, max_img_loss, i, net.max_batches, mean_average_precision, draw_precision);
    }
    #endif // OPENCV
}
```

**Comment by AlexeyAB** [comment]
> One other issue though, currently mAP is ignored unless you compile with opencv. This isn't necessary. I don't plot the results, I just direct stdout to a file and parse that later. I modified the code to something like this which works fine:

Yes, it is ok.


> Thanks, I think it's just NMS being crushed for a huge number of boxes. I found (for my dataset) after 500 iterations it's fine, but I wanted to capture the initial changes in mAP and 1000 was too late. I think as I'm training on a single fairly easy class the network is converging very quickly.

Do you mean that you got the highest possible mAP before 1000 iterations?
What model do you use and do you use burn_in=1000?

Also just to know, on practice you can't overfit Yolo v3. (Overfitting of Yolo v3 possible only in theory)

**Comment by jveitchmichaelis** [comment]
It's yolov3, standard configuration with custom anchors (and 640x512) and 1000 iteration burn in time.

Here's the loss/map curve. From looking through the dataset, it definitely looks like it's overfitting the training data. Maybe there are too many self-similar images or not enough augmentation. It's a little after 1000, more like 1600-1800.

![loss curve](https://i.ibb.co/3BYpYJJ/loss.png)

**Comment by AlexeyAB** [comment]
* So did you train modified `yolov3.cfg` about `5*34000/64 =~ 2500` iterations?
* Try to train at least 6000-8000 iterations, will mAP increase after 2500 iterations?
*  Can you shor mAP chart for 16 epochs (530 iterations x 16 epochs ~= 8500 iterations)?

There is too low number of iterations to say that there is overfitting. May be it is just accuracy fluctuation.

If there is still overfitting starting from 3.5 epochs (~2000 iterations), even if you train 16 epochs (8500 iterations). Then try to set `steps=2000, 2500` and train starting from 2000 iterations, mAP should increase.


![image](https://user-images.githubusercontent.com/4096485/50699681-976abf80-1059-11e9-85d3-1f141c706731.png)

**Comment by jveitchmichaelis** [comment]
Here's a similar result from the same dataset but taking every third image:

![](https://i.ibb.co/rkTfw61/loss-split.png)

After a while the mAP flatlined and I didn't see much point wasting time training further, since there are other probably better ways to improve the model robustness.

**Comment by AlexeyAB** [comment]
@jveitchmichaelis When you see that mAP doesn't increase more, you can try to decrease 10x learning_rate and train more. Then again, when mAP doesn't increase, set ecrease 10x learning_rate. Usually you can get +1% mAP.

Or just set correct values `steps=10000, 12000`, and learning rate will be decrease automatically at 10000 iterations (20 Epochs) and 12000 iterations.
Or `steps=2000, 2500` if you are sure that mAP doesn't increase more.

![image](https://user-images.githubusercontent.com/4096485/50724033-8a9ca900-10f7-11e9-8e0c-cb719b6da550.png)

----
